### ISSUE TITLE
Integrate with Lance

### ISSUE THREAD

**Issue by xloya** [NVIDIA/Megatron-Energon#153]
[Lance](https://github.com/lancedb/lance) is an open table and file format that is easy and fast to version, query and train on. It's designed to be used with images, videos, 3D point clouds, audio and of course tabular data. It supports any POSIX file systems, and cloud storage like AWS S3 and Google Cloud Storage. The key features of Lance include:

High-Performance Random Access: 100x faster than Parquet.

Zero-Copy Data Evolution: add, drop or update column data in existing rows without the need to rewrite the entire dataset.

Multimodal Data: natively store large text, images, videos, documents and embeddings all together, and efficiently access them using the Blob API.

Vector Search: find nearest neighbors in under 1 millisecond with IVF-PQ, IVF-SQ, HNSW and more.

Full-Text Search: fast search over text with inverted index, Ngram index plus different tokenizers.

Hybrid Search: perform hybrid vector, full-text and OLAP queries all using the same format.

Row Level Transaction: fully ACID transaction with row level conflict resolution.

Ecosystem Integrations: PyTorch, Tensorflow, Ray, HuggingFace, Apache Arrow, Apache DataFusion, Apache Spark, Trino, DuckDB, PostgreSQL, and more on the way.

Compared with WebDataset, Lance has better random reading and point query capabilities, which significantly improves the efficiency of model training. Is Megatron-Energon considering supporting the integration of Lance?

**Comment by xloya** [comment]
> Hi [@xloya](https://github.com/xloya), Energon is very much built on and optimized for the WebDataset format. We are not planning to integrate other formats at the moment.
> 
> Note however, that you can create a polylithic energon dataset with lance if you like to. I.e. you create a small primary WebDataset (with just text data and pointers) and then in your custom cooker you read the corresponding entry from the lance file:
> 
> See https://nvidia.github.io/Megatron-Energon/advanced/crude_datasets.html
> 
> I'd be surprised though, if lance is faster than our WebDataset-based format, since we are using direct binary indices and making sure to employ consecutive reads for fast reading.

Thanks for your reply. We currently use Megatron and WebDataset to support model training, and it works well. In some public practices of Lance, there is indeed a significant improvement compared to the native WebDataset, but I am not sure whether it has a clear advantage compared to the deep optimization of WebDataset by Megatron-Energon, because there is currently no way for us to directly compare the performance of the two when using Megatron.
